nixos/modules/services/monitoring/prometheus/collectd-exporter.nix

```diff
@@ -0,0 +1,134 @@
+        type = types.nullOr types.str;
+        default = null;
+        example = "0.0.0.0";
+        description = ''
+          Address to listen on for web interface, telemetry and collectd JSON data.
+        '';
+      };
+
+      collectdBinary = mkOption {
+        description = ''
+          Configure collection via the collectd binary network protocol.
+        '';
+        type = (types.submodule {
+          options = {
+            enable = mkEnableOption "collectd binary protocol receiver";
+
+            authFile = mkOption {
+              default = null;
+              type = types.nullOr types.path;
+              description = "File mapping user names to pre-shared keys (passwords).";
+            };
+
+            port = mkOption {
+              type = types.int;
+              default = 25826;
+              description = ''Network address on which to accept collectd binary network packets.'';
+            };
+
+            listenAddress = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "0.0.0.0";
+              description = ''
+                Address to listen on for binary network packets.
+              '';
+            };
+
+            securityLevel = mkOption {
+              type = types.enum ["None" "Sign" "Encrypt"];
+              default = "None";
+              description = ''
+                Minimum required security level for accepted packets.
+              '';
+            };
+          };
+        });
+      };
+
+      extraFlags = mkOption {
+        type = types.listOf types.str;
+        default = [];
+        description = ''
+          Extra commandline options when launching the collectd exporter.
+        '';
+      };
+
+      logFormat = mkOption {
+        type = types.str;
+        default = "logger:stderr";
+        example = "logger:syslog?appname=bob&local=7 or logger:stdout?json=true";
+        description = ''
+          Set the log target and format.
+        '';
+      };
+
+      logLevel = mkOption {
+        type = types.enum ["debug" "info" "warn" "error" "fatal"];
+        default = "info";
+        description = ''
+          Only log messages with the given severity or above.
+        '';
+      };
+
+      openFirewall = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Open port in firewall for incoming connections.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    networking.firewall.allowedTCPPorts = (optional cfg.openFirewall cfg.port) ++ (optional (cfg.openFirewall && cfg.collectdBinary.enable) cfg.collectdBinary.port);
+
+    systemd.services.prometheus-collectd-exporter = {
+      description = "Prometheus exporter for Collectd metrics";
+      unitConfig.Documentation = "https://github.com/prometheus/collectd_exporter";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        User = "nobody";
```

Comment:
How about using `DynamicUser` , which we have in systemd since some versions?

Reply:
I didn't know about that. I just copied it from the other collectors. A quick tests shows it seems to work. I updated the PR accordingly.

Reply:
Seems like this would be the first module in nixpkgs using `DynamicUser`.

Reply:
Many services have persistent data, so that DynamicUser is not applicable. Also DynamicUser is still not well known and new.